secure/net/src/grpc_network_service/mod.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::network_controller::{Message, MessageType};
+use aptos_logger::{error, info};
+use aptos_protos::remote_executor::v1::{
+    remote_execution_client::RemoteExecutionClient,
+    remote_execution_server::{RemoteExecution, RemoteExecutionServer},
+    Empty, NetworkMessage, FILE_DESCRIPTOR_SET,
+};
+use crossbeam_channel::Sender;
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{Arc, Mutex},
+};
+use tokio::{runtime::Runtime, sync::oneshot};
+use tonic::{
+    transport::{Channel, Server},
+    Request, Response, Status,
+};
+
+pub struct RemoteExecutionServerWrapper {
+    inbound_handlers: Arc<Mutex<HashMap<MessageType, Sender<Message>>>>,
+}
+
+impl RemoteExecutionServerWrapper {
+    pub fn new(inbound_handlers: Arc<Mutex<HashMap<MessageType, Sender<Message>>>>) -> Self {
+        Self { inbound_handlers }
+    }
+
+    // Note: The object is consumed here. That is once the server is started, we cannot/should not
+    //       use the object anymore
+    pub fn start(
+        self,
+        rt: &Runtime,
+        _service: String,
+        server_addr: SocketAddr,
+        server_shutdown_rx: oneshot::Receiver<()>,
+    ) {
+        rt.spawn(async move {
+            self.start_async(server_addr, server_shutdown_rx).await;
+        });
```

Comment:
Yes, one_shot channel is a good idea.. This is basically my  3 of my TODOs. Waiting for server to start before sending messages. Hence those sleeps in test cases

Reply:
There is no easy way to do this by APIs provided by GRPC or Tonic. Ways to achieve this would be by doing a check/retry from client side.
As discussed going to leave this a TODO for now

Reply:
One way to do this would be implementing a healthcheck service
https://stackoverflow.com/questions/66798278/golang-grpc-how-to-determine-when-the-server-has-started-listening

Reply:
Health check service is a good idea.

Reply:
Maybe we can just use the reflection service?

Reply:
That is a good point, I will check into it